File: apps/android/app/src/main/kotlin/com/nogeass/passalarm/domain/usecase/RescheduleNextNUseCase.kt
package com.nogeass.passalarm.domain.usecase

import com.nogeass.passalarm.domain.model.ScheduledToken
import com.nogeass.passalarm.domain.model.TokenStatus
import com.nogeass.passalarm.domain.repository.ScheduledTokenRepository
import com.nogeass.passalarm.domain.scheduler.AlarmScheduler
import javax.inject.Inject

class RescheduleNextNUseCase @Inject constructor(
    private val computeQueue: ComputeQueueUseCase,
    private val tokenRepository: ScheduledTokenRepository,
    private val scheduler: AlarmScheduler
) {
    companion object {
        const val MAX_TOKENS = 60
    }

    suspend operator fun invoke() {
        // 1. Cancel existing
        val existing = tokenRepository.fetchPending()
        existing.forEach { scheduler.cancelAlarm(it) }
        tokenRepository.deletePending()

        // 2. Compute fresh merged queue from all enabled plans
        val queue = computeQueue.execute()
        val active = queue.filter { !it.isSkipped }.take(MAX_TOKENS)

        // 3. Schedule new tokens
        val tokens = active.map { occurrence ->
            val osId = occurrence.planId.toInt() * 100_000_000 +
                    occurrence.date.replace("-", "").toInt() % 100_000 * 1000 +
                    occurrence.timeHHmm.replace(":", "").toInt()
            ScheduledToken(
                planId = occurrence.planId,
                date = occurrence.date,
                fireAtEpoch = occurrence.fireAtEpoch,
                osIdentifier = osId,
                status = TokenStatus.PENDING
            )
        }

        tokens.forEach { token ->
            val savedId = tokenRepository.save(token)
            val savedToken = token.copy(id = savedId)
            scheduler.scheduleAlarm(savedToken)
        }
    }
}

File: apps/ios/PassAlarm/Domain/UseCase/RescheduleNextNUseCase.swift
import Foundation

final class RescheduleNextNUseCase: Sendable {
    private let computeQueue: ComputeQueueUseCase
    private let tokenRepository: ScheduledTokenRepositoryProtocol
    private let scheduler: NotificationSchedulerProtocol

    static let scheduleCount = 60

    init(computeQueue: ComputeQueueUseCase,
         tokenRepository: ScheduledTokenRepositoryProtocol,
         scheduler: NotificationSchedulerProtocol) {
        self.computeQueue = computeQueue
        self.tokenRepository = tokenRepository
        self.scheduler = scheduler
    }

    func execute() async throws {
        // 1. Cancel all existing scheduled tokens
        let existing = try await tokenRepository.fetchScheduled()
        let existingIds = existing.map { $0.osIdentifier }
        if !existingIds.isEmpty {
            try await scheduler.cancel(identifiers: existingIds)
        }
        try await tokenRepository.deleteScheduled()

        // 2. Compute fresh queue
        let queue = try await computeQueue.execute()
        let activeOccurrences = queue.filter { !$0.isSkipped }
        let topN = Array(activeOccurrences.prefix(Self.scheduleCount))

        // 3. Schedule new tokens
        var tokens: [ScheduledToken] = []
        for occurrence in topN {
            let planPrefix = occurrence.planId.uuidString.prefix(8)
            let timeCompact = occurrence.timeHHmm.replacingOccurrences(of: ":", with: "")
            let osId = "passalarm-\(planPrefix)-\(occurrence.date)-\(timeCompact)"
            let token = ScheduledToken(
                id: UUID(),
                planId: occurrence.planId,
                date: occurrence.date,
                fireAtEpoch: occurrence.fireDate.timeIntervalSince1970,
                osIdentifier: osId,
                status: .scheduled,
                createdAt: Date(),
                updatedAt: Date()
            )
            tokens.append(token)

            try await scheduler.schedule(
                identifier: osId,
                at: occurrence.fireDate,
                title: "パスアラーム",
                body: occurrence.planLabel.isEmpty
                    ? "\(occurrence.timeHHmm) アラーム"
                    : "\(occurrence.planLabel) \(occurrence.timeHHmm)",
                soundId: occurrence.soundId
            )
        }

        try await tokenRepository.saveAll(tokens)
    }
}
